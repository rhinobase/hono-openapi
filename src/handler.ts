import type { Context, Env, Hono } from "hono";
import type {
  BlankEnv,
  BlankInput,
  BlankSchema,
  Input,
  Schema,
} from "hono/types";
import type { OpenAPIV3_1 } from "openapi-types";
import type {
  DescribeRouteOptions,
  GenerateSpecOptions,
  ResolverReturnType,
  SanitizedGenerateSpecOptions,
} from "./types";
import {
  ALLOWED_METHODS,
  type AllowedMethods,
  registerSchemaPath,
  removeExcludedPaths,
  uniqueSymbol,
} from "./utils";

const DEFAULT_OPTIONS: Partial<GenerateSpecOptions> = {
  documentation: {},
  excludeStaticFile: true,
  exclude: [],
  excludeMethods: ["OPTIONS"],
  excludeTags: [],
};

type SpecContext = {
  components: Record<string, OpenAPIV3_1.SchemaObject>;
  options: SanitizedGenerateSpecOptions;
};

/**
 * Generate OpenAPI specs for the given Hono instance
 * @param hono Instance of Hono
 * @param options Options for generating OpenAPI specs
 * @param config Configuration for OpenAPI route handler
 * @param Context Route context for hiding routes
 * @returns OpenAPI specs
 */
export async function generateSpecs<
  E extends Env = BlankEnv,
  P extends string = string,
  I extends Input = BlankInput,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  options = DEFAULT_OPTIONS,
  c?: Context<E, P, I>,
) {
  const ctx: SpecContext = {
    components: {},
    // @ts-expect-error
    options: {
      ...DEFAULT_OPTIONS,
      ...options,
    },
  };

  const _documentation = ctx.options.documentation ?? {};
  const schema = await generatePaths(hono, ctx);

  // Hide routes
  for (const path in schema) {
    for (const method in schema[path]) {
      const isHidden = getHiddenValue({
        valueOrFunc: schema[path][method]?.hide,
        method,
        path,
        c,
      });

      if (isHidden) {
        delete schema[path][method];
      }
    }
  }

  return {
    openapi: "3.1.0",
    ..._documentation,
    tags: _documentation.tags?.filter(
      (tag) => !ctx.options.excludeTags?.includes(tag?.name),
    ),
    info: {
      title: "Hono Documentation",
      description: "Development documentation",
      version: "0.0.0",
      ..._documentation.info,
    },
    paths: {
      ...removeExcludedPaths(schema, ctx),
      ..._documentation.paths,
    },
    components: {
      ..._documentation.components,
      schemas: {
        ...ctx.components,
        ..._documentation.components?.schemas,
      },
    },
  } satisfies OpenAPIV3_1.Document;
}

async function generatePaths<
  E extends Env = BlankEnv,
  P extends string = string,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  ctx: SpecContext,
): Promise<OpenAPIV3_1.PathsObject> {
  const paths: OpenAPIV3_1.PathsObject = {};

  for (const route of hono.routes) {
    // Finding routes with uniqueSymbol
    if (!(uniqueSymbol in route.handler)) {
      // Include empty paths, if enabled
      if (ctx.options.includeEmptyPaths) {
        registerSchemaPath({
          route,
          paths,
        });
      }

      continue;
    }

    const routeMethod = route.method as AllowedMethods | "ALL";

    // All method acts like a middleware, so we can skip it
    if (routeMethod !== "ALL") {
      // Exclude methods
      if (ctx.options.excludeMethods?.includes(routeMethod)) {
        continue;
      }

      // Include only allowed methods
      if (!ALLOWED_METHODS.includes(routeMethod)) {
        continue;
      }
    }

    const middlewareHandler = route.handler[uniqueSymbol] as
      | ResolverReturnType
      | { schema: DescribeRouteOptions; components: undefined };

    const defaultOptionsForThisMethod = ctx.options.defaultOptions
      ?.[routeMethod];

    const { schema: routeSpecs, components = {} } =
      "schema" in middlewareHandler
        ? middlewareHandler
        : await middlewareHandler
          .toOpenAPISchema({
            ...defaultOptionsForThisMethod,
            components: ctx.components,
          });

    ctx.components = {
      ...ctx.components,
      ...components,
    };

    registerSchemaPath({
      route,
      specs: routeSpecs,
      paths,
    });
  }

  return paths;
}

function getHiddenValue(options: {
  valueOrFunc: DescribeRouteOptions["hide"];
  c?: Context;
  method: string;
  path: string;
}) {
  const { valueOrFunc, c, method, path } = options;

  if (valueOrFunc != null) {
    if (typeof valueOrFunc === "boolean") {
      return valueOrFunc;
    } else if (typeof valueOrFunc === "function") {
      if (c) {
        return valueOrFunc(c);
      } else {
        console.warn(
          `'c' is not defined, cannot evaluate hide function for ${method} ${path}`,
        );
      }
    }
  }

  return false;
}
