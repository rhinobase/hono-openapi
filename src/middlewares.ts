import { type Hook, sValidator } from "@hono/standard-validator";
import type { Env, Input, MiddlewareHandler, ValidationTargets } from "hono";
import type { StandardSchemaV1 } from "@standard-schema/spec";
import { toJsonSchema } from "@standard-community/standard-json";
import { toOpenAPISchema } from "@standard-community/standard-openapi";
import { uniqueSymbol } from "./utils";
import { DescribeRouteOptions } from "./types";

/**
 * Generate a resolver for a validation schema
 * @param schema Validation schema
 * @returns Resolver result
 */
export function resolver<Schema extends StandardSchemaV1>(schema: Schema) {
  return {
    vendor: schema["~standard"].vendor,
    validate: schema["~standard"].validate,
    toJSONSchema: (options?: Record<string, unknown>) =>
      toJsonSchema(schema, options),
    toOpenAPI: (options?: Record<string, unknown>) =>
      toOpenAPISchema(schema, options),
  };
}

type HasUndefined<T> = undefined extends T ? true : false;

/**
 * Create a validator middleware
 * @param target Target for validation
 * @param schema Validation schema
 * @param hook Hook for validation
 * @returns Middleware handler
 */
export function validator<
  Schema extends StandardSchemaV1,
  Target extends keyof ValidationTargets,
  E extends Env,
  P extends string,
  In = StandardSchemaV1.InferInput<Schema>,
  Out = StandardSchemaV1.InferOutput<Schema>,
  I extends Input = {
    in: HasUndefined<In> extends true ? {
        [K in Target]?: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]?: ValidationTargets[K][K2] };
      }
      : {
        [K in Target]: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]: ValidationTargets[K][K2] };
      };
    out: { [K in Target]: Out };
  },
  V extends I = I,
>(
  target: Target,
  schema: Schema,
  hook?: Hook<StandardSchemaV1.InferOutput<Schema>, E, P, Target>,
  options?: Record<string, unknown>,
): MiddlewareHandler<E, P, V> {
  const middleware = sValidator(target, schema, hook);

  // @ts-expect-error not typed well
  return Object.assign(middleware, {
    [uniqueSymbol]: {
      target,
      ...resolver(schema),
      options,
    },
  });
}

/**
 * Describe a route with OpenAPI specs.
 * @param specs Options for describing a route
 * @returns Middleware handler
 */
export function describeRoute(specs: DescribeRouteOptions): MiddlewareHandler {
  const middleware: MiddlewareHandler = async (_c, next) => {
    await next();
  };

  return Object.assign(middleware, {
    [uniqueSymbol]: {
      specs,
    },
  });
}
