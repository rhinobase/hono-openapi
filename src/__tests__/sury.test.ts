import { Hono } from "hono";
import * as S from "sury";
import { describe, expect, it } from "vitest";
import { generateSpecs } from "../handler.js";
import { describeRoute, resolver, validator } from "../middlewares.js";

describe("sury", () => {
  it("basic", async () => {
    const app = new Hono().get(
      "/",
      describeRoute({
        tags: ["test"],
        summary: "Test route",
        description: "This is a test route",
        responses: {
          200: {
            description: "Success",
            content: {
              "application/json": {
                schema: resolver(
                  S.schema({
                    message: S.string,
                  }),
                ),
              },
            },
          },
        },
      }),
      validator("json", S.schema({ message: S.string })),
      async (c) => {
        return c.json({ message: "Hello, world!" });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });

  it("with metadata", async () => {
    const app = new Hono().get(
      "/",
      describeRoute({
        tags: ["test"],
        summary: "Test route",
        description: "This is a test route",
        responses: {
          200: {
            description: "Success",
            content: {
              "application/json": {
                schema: resolver(
                  S.extendJSONSchema(
                    S.schema({
                      myString: S.string,
                      myUnion: S.union([S.number, S.boolean]),
                    }),
                    { $id: "SuccessResponse" },
                  ),
                ),
              },
            },
          },
        },
      }),
      validator("json", S.schema({ message: S.string })),
      async (c) => {
        return c.json({ message: "Hello, world!" });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });
});
