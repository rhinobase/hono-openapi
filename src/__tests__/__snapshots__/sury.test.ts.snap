// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sury > basic 1`] = `
{
  "components": {},
  "info": {
    "description": "Development documentation",
    "title": "Hono Documentation",
    "version": "0.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "description": "This is a test route",
        "operationId": "getIndex",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": true,
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "message",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": true,
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "summary": "Test route",
        "tags": [
          "test",
        ],
      },
    },
  },
  "tags": undefined,
}
`;

exports[`sury > with metadata 1`] = `
{
  "components": {
    "schemas": {
      "SuccessResponse": {
        "additionalProperties": true,
        "properties": {
          "myString": {
            "type": "string",
          },
          "myUnion": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "type": "boolean",
              },
            ],
          },
        },
        "required": [
          "myString",
          "myUnion",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "Development documentation",
    "title": "Hono Documentation",
    "version": "0.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "description": "This is a test route",
        "operationId": "getIndex",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": true,
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "message",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SuccessResponse",
                },
              },
            },
            "description": "Success",
          },
        },
        "summary": "Test route",
        "tags": [
          "test",
        ],
      },
    },
  },
  "tags": undefined,
}
`;
